### ISSUE TITLE
Wundefined-func-template warns on explicit instantiations: false positive?

### ISSUE THREAD

**Issue by carlosgalvezp** [llvm/llvm-project#52653]
Hi!

If I enable the warning "Wundefined-func-template" it triggers in the following valid code:

```
// foo.h
#pragma once

template <typename T>
void foo();
```
```
// foo.cpp
#include "foo.h"

template <typename T>
void foo() {}

template void foo<int>();
```
```
// main.cpp
#include "foo.h"

int main() {
  foo<int>();
}
```

Compiling, I get:

```
 clang++ -Wundefined-func-template main.cpp foo.cpp 
main.cpp:4:3: warning: instantiation of function 'foo<int>' required here, but no definition is available [-Wundefined-func-template]
  foo<int>();
  ^
./foo.h:4:6: note: forward declaration of template entity is here
void foo();
     ^
main.cpp:4:3: note: add an explicit instantiation declaration to suppress this warning if 'foo<int>' is explicitly instantiated in another translation unit
  foo<int>();
  ^
1 warning generated.
```

What's the problem? I have indeed added an explicit instantiation in `foo.cpp` and the code compiles and runs fine. Is this a FP or am I missing something?

Thanks!

**Comment by zero9178** [comment]
The compiler is complaining as when it is compiling main.cpp it does not know that another file is instantiating the function, and therefore tries to instantiate it, but fails as no definition is visible. Hence the warning. From the perspective of clang compiling main.cpp, you just happen to be lucky that another file does instantiate it.

The way to fix this warning as well as the C++ standard way to inform the compiler not to attempt to instantiate a template unless explicitly specified is via putting a `extern template` declaration in the header. See also this SO post that happens to be talking about the same scenario  https://stackoverflow.com/questions/8130602/using-extern-template-c11
I'd recommend putting the declaration in the header and keeping the explicit instantiation in your foo.cpp file

**Comment by carlosgalvezp** [comment]
Aha, thanks for the clarification! So the header would look like this?

```
// foo.h
template <typename T>
void foo();

extern template void foo<int>();
```

I suppose `extern` is really key here, otherwise it would look identical to the explicit instantiation in the .cpp file.
